Article first draft (#135)
hassanobeid1994/tr_b_causal_2020: article/sections/_4_graph_construction.tex

\section{(Initial) Causal Graph Construction}
\label{sec:graph-construction}

Construction of an initial causal graph typically proceeds as follows.
At a high level, we
\begin{enumerate}
   \item adopt a population perspective,
   \item brainstorm all variables that we think affect the system that generates our observations,

>    \item remove any variables that would cause bias in our causal inferences,

Reviewer: ```suggestion
   \item remove any variables that could cause bias in our causal inferences,
```

>    \item connect all variables in our graph according to our a-priori beliefs about causal relations amongst these variables,

Reviewer: ```suggestion
   \item connect all variables in our graph according to our a-priori beliefs about causal relations amongst them,
```

>    \item consider how the graph structure may differ between individuals and subgroups within the population.

Reviewer: What could this mean in terms of practical steps on a high level? Does it mean drawing other plausible graphs for specific subgroups? I know you go into this in detail later in the section but I think it would be good to give a high level example here.

Author: Yep, that's exactly what it means! How do you think I should word this to be both clear AND non-repetitive?

For instance, I think the phrase "differing graph structures between subgroups" literally means differing graphs for differing subgroups. How  can I say this more clearly?

My first idea, "consider drawing different graphs for differing subgroups and individuals within the population," sounds very intimidating and crazy to me. I.e., drawing one graph is hard, but now you want me to draw many? Hopefully I can balance all these considerations!

\end{enumerate}

> The following paragraphs will describe these steps in detail.

Reviewer: ```suggestion
The following paragraphs describe these steps in detail.
```

To begin, we adopt the position of a researcher concerned about population level relationships.
This means we will think through what is a likely generative model for all individuals.

> Later on, we will devote time to thinking about how subgroups and individual heterogeneity may affect our causal graphs.

Reviewer: ```suggestion
After creating a plausible generative model for all individuals, we will devote time to thinking about how subgroups and individual heterogeneity may affect our causal graphs and potentially result in different causal graphs.
```

Author: Similar question as  elsewhere: what sort of affects from subgroups and individual heterogeneity are you thinking of that don't result in a different causal graph?

Also, I changed the beginning of the sentence to say "Later in this section [...]". I didn't want to go with the suggested edit because it makes the sentence preamble appreciably longer with large words, and it would divert the sentence's focus from what will go on later to more of what happened before.

> Now, we add our first variables to our graph, the outcome variable(s) of interest.

Reviewer: ```suggestion
Now, we add our first variable(s) to our graph, the outcome variable(s) of interest in our problem.
```

> Then, we note all the variables that we believes to influence the variable(s) of interest.

Reviewer: ```suggestion
Then, we list all the variables we believe to influence the variable(s) of interest.
```

> We refer to these variables to as our initial explanatory variables.

Reviewer: ```suggestion
We refer to these variables as our initial explanatory variables.
```

> Next, we iterate through the initial explanatory variables.

Reviewer: I think given the different possible types of explanatory variables, we should define them here as opposed to during the detailed descriptions of how to add them to the causal graph.

Author: Why do you think this will be better?

I think we'll end up with a paragraph full of nothing but definitions. My guess is that this paragraph will be:
- hard to read because it's made of almost purely technical definitions
- lacking contextual information that appears in the paragraph alongside the introduction of each variable type. If I put all the definitions together, then I think I'd have to leave this contextual information for the paragraphs after the definitions. I think the see-sawing between talking about a type of variable then going away then returning to give some more info about it would be unpleasant to read.

That said, I'll switch with a quickness if that seems like it'd be most useful for readers! Let me know what you think in light of the info above.

Reviewer: It's just a preference thing. I usually like to tell people what I'll tell them about, then tell them about it. But from our convo two weeks ago, I think it's fair to keep it the way it is.

For each current explanatory variable in the iteration, we think of variables that may modify the effect of the current explanatory variable on the outcome(s) of interest.
We refer to these variables as effect modifiers.
Note that some effect modifiers may be a part of our initial explanatory variables.

> For any effect modifiers that one thinks of, outside of the initial explanatory variables, add them to our causal graph.

Reviewer: ```suggestion
Next, we iterate through the determined initial explanatory variables.
For each current explanatory variable in the iteration, we think of variables that may modify the effect of the current explanatory variable on the outcome(s) of interest.
We refer to these variables as effect modifiers.
Note that some effect modifiers may be a part of our list of initial explanatory variables.
For any effect modifiers that we think of, outside of the list of initial explanatory variables, we add them to our causal graph.
```

After adding explanatory and effect modifying variables to the graph, we turn our attention to mediating variables.
Again, we iterate through each explanatory variable.
On each iteration, we brainstorm all the variables through which our current explanatory variable influences the outcome.
For instance, consider how the presence of a bike lane influences bicycle mode choice.
We hypothesize that an individual's subjective perception of safety is be the primary (or sole) variable through which bicycle lane presence influences mode choice.

> Accordingly, we add subjective perception of safety to our causal graph for travel mode choice.

Reviewer: ```suggestion
After adding explanatory and effect modifying variables to the graph, we turn our attention to mediating variables. A mediating variable is one through which an explanatory variable influences the outcome(s) of interest.
Again, we iterate through each explanatory variable.
On each iteration, we brainstorm all possible mediating variables.
For instance, consider how the presence of a bike lane influences bicycle mode choice.
We hypothesize that an individual's subjective perception of safety is be the primary (or sole) variable through which bicycle lane presence influences mode choice.
Accordingly, we add subjective perception of safety as a mediating variable to our causal graph for travel mode choice.
```

Coming to our second to last category of variables, we think of confounding variables.
The process is similar to how we generated effect modifying variables.
We iterate through each of the explanatory, mediating, and effect modifying variables, thinking specifically of any variables that both cause the current variable in the iteration and cause the outcome variable(s).
We call these variables, which cause our outcome and current variables in the iteration, confounding variables.

> We will add each of them to our causal graph.

Reviewer: Can we give an example here similarly to what we did for mediating variables?

For the last set of variables, we should explicitly consider the role of time, even in research that may be cross-sectional due to the data that is available to us.
In reality, how do we think our system evolves over time?
If we consider multiple observations of a given decision maker, how do variables observed of that decision maker at time $t$ partially cause future variables important to the context or outcomes for that decision maker at time $t+1$?
How do the actions of a decision maker $i$ at time $t$ partially cause the future context or outcomes of a decision maker $j$?

> We should add explicit nodes to our graph, subscripted or denoted by time, to show the cross-time causal relationships in our system.

Reviewer: ```suggestion
For the last set of variables, we should explicitly consider the role of time, even in research that may be cross-sectional due to the data that is available to us or the context of the research itself.
In reality, how do we think our system evolves over time?
For example, if we consider multiple observations of a given decision maker, how do observed variables of that decision maker at time $t$ partially cause or affect future variables important to the context or outcome(s) of interest for that decision maker at time $t+1$ or later?
How do the actions of a decision maker $i$ at time $t$ partially cause the future context or outcomes of a decision maker $j$ (either at time $t$ or later)?
We should add explicit nodes to our graph, subscripted or denoted by time, to show the cross-time causal relationships in our system.
```

Author: Hmm, for line 45, doesn't the "future context" specify the reference to a later time? The idea that $i$'s actions cause $j$'s decision within the current time is most closely related to the area of network interference within causal inference.

Now that you mention it, I realize I've neglected any explicit mention of network interference (mostly because I haven't done any causal inference work on the topic and it isn't foremost in my mind).

Nevertheless, I still don't think this paragraph is the best place for that. Interference within the current time would show up as a node on one individual's graph that happens to be the outcome of another individual. I don't think there's anything special about that other individual's outcome node, save maybe an undirected edge to the current individual's outcome node. It can't be directed since they may both happen and affect each other at the same time. All of the special treatment happens in the analysis of data subject to network interference, not in the graph construction.

That said, I should def add a reference to the topic, likely in the "catch-all" discussion of causal inference and transportation choice modelling that didn't make it into the body of this chapter.

Reviewer: I think what I should've done is delete "future context" and keep my added suggestion there because it accounts for network interference and future context.

At this point, we have added all the outcome, explanatory, effect modifying, mediating, confounding, and time-indexed variables that we can think of to our causal graph.
They are all disconnected nodes, singletons, in the graph.
We now focus on pruning nodes from this graph, before drawing our final hypothesized connections.
In particular, we focus on pruning ``post-outcome'' variables that are not part of the causal graph for future time periods or other observations.

> The reason for this is that conditioning on such post-outcome variables biases one's causal effect estimates.

Reviewer: ```suggestion
At this point, we have added all the outcome, explanatory, effect modifying, mediating, confounding, and time-indexed variables that we believe are relevant to our context to our causal graph.
However, at this time they are all disconnected nodes, singletons, in the graph.
We now focus on pruning nodes from this graph, before drawing our final hypothesized connections.
In particular, we focus on pruning ``post-outcome'' variables that are not part of the causal graph for future time periods or other observations.
The reason for this is that conditioning on such post-outcome variables would bias our causal effect estimates.
```

> To remove the problematic variables, we iterate through each of the non-outcome variables in our graph, and we assess whether the variable is actually a result of the outcome (perhaps in combination with other variables in our graph).
> These post-outcome variables temporally follow the outcome variable(s) but do not cause variables in the causal graph for other observations.
> We remove all such post-outcome variables from our graph.

Reviewer: ```suggestion
To remove the problematic variables, we iterate through each of the non-outcome variables in our graph, and we assess whether a variable is actually a result of the outcome (perhaps in combination with other variables in our graph).
These post-outcome variables temporally follow the outcome variable(s) but do not cause or affect variables in the causal graph for other observations.
We remove all such post-outcome variables from our graph.
```

Author: What sort of affects do you have in mind? I think of affect and cause as synonyms in this case, and I'm not immediately picturing a way that variable A affects variable B without causing variable B.

For clarity, I'm definitely thinking of an affect on say the distribution of B to be included when I say A causes B, even if the actual outcome is the same. Imagine A and B are both binary. If I change the probability of B =1, even if a certain individual's outcome is actually B = 0 under A = 1 and A = 0, then I still say A caused B because A had an effect on what B could have been.

Reviewer: My suggestion was because when I use "cause" I use it as "cause a change in a variable" as opposed to cause a variable. It's just how we're reading things, not so much the interpretation of how post-outcome variables affect variables in causal graphs for other observations.

> Now is a good time to step back and consider what other researchers have thought.
> Specifically, we should conduct a literature review to see how other researchers have conceptualized the problem that we are working on.
> Have they included variables that we have not?
> Were those variables related our outcomes of interest?
> If so, should we add these variables to our causal graph, and how?
> Do the included variables of other researchers point suggest the existence of confounders in their work that we should include in our graph?
> Have other researchers ascribed differing roles to our graph's current variables than we have?
> For example, have other researchers judged a variable to be confounder, when we solely thought of the variable as an affect modifier?
> Critically examine the evidence for these alternative decisions to see if we should also reconsider how we're judging our variables.

Reviewer: ```suggestion
Now is a good time to step back and consider what other researchers have thought about the problem of interest.
Specifically, we should conduct a literature review to see how other researchers have conceptualized the problem that we are working on.
Have they constructed causal graphs to clearly represent the assumptions of their analysis?
Have they considered or included variables that we have not?
Were those variables related to our outcomes of interest?
If so, should we add these variables to our causal graph?
How should such variables enter our causal graph?
Do the included variables of other researchers suggest the existence of confounders in their work that we should include in our graph?
Have other researchers ascribed differing roles to our graph's current variables than we have?
For example, have other researchers judged a variable to be confounder, when we solely thought of the variable as an affect modifier?
As we answer these questions, and any others that may arise as a result of exploring the literature, we should critically examine the evidence for these alternative decisions to see if we should also reconsider how we are judging our variables and including them in our graph.
```

Author: Mo! Question about line 60: "Have they constructed causal graphs to clearly represent the assumptions of their analysis?" After considering the other questions in the list, what do we learn specifically from knowing that they constructed a causal graph for their analysis?

My hope with the list was to capture the common things one would want to learn from examining the causal graph from others, if they have one. Then if I specifically added those questions, I wouldn't need to necessarily be concerned with whether the papers I'm reading construct a causal graph. If I abstract/extract the core of the questions, then they should hopefully be answerable even if a researcher is operating from a potential outcomes framework and didn't draw any causal graph.

Reviewer: Well the reason I had in there is to include cases/studies where other researchers haven't even bothered to put together causal graphs at all.

Finally, we need to connect the variables in our graph.
\begin{enumerate}
   \item Draw direct arrows from our explanatory variables, confounders, and effect modifiers to the outcomes.
   \item Draw arrows from the explanatory variables to the mediators, and then draw arrows from the mediators to the outcomes.
   \item Draw arrows from the confounders to the explanatory variables and mediators that they may cause.
   \item Draw arrows from the variables in time $t$ to the variables that they cause in time $t+1$.
\end{enumerate}
After drawing in all arrows, we should now have a fully connected causal graph.
Take a moment to look over the graph to ensure there are no remaining singletons and that we have not drawn any spurious connections.
Also, take a moment to celebrate.

> Drawing one's first causal graph is hard work!

Reviewer: The switch from "we" to talking directly to the reader is a bit weird. Should we stay consistent?

Author: Weird in the "oh snap, they're talking to ME!" sense or weird in the "whoa, this shouldn't happen sense?

I actually want the first kind of weird. I literally want the reader to pause and reflect a bit. I've edited this in the first revision to actually include a sentence that just says "[p]ause." Hopefully that will be enough of a visual cue that something different is going on, followed up by the linguistic cue in switching from point-of-view from us to the reader.

If you felt the second kind of weird, I'd definitely want to know why!

Reviewer: It's the first type. Got it if that's the intent.

> After celebrating, take a moment to pursue the following graph editing exercises.
> First, think about how the graph might differ across sub-populations.
> What sub-populations exist in your population of interest?
> Are there any causal relationships that should not exist for a given sub-population?
> For instance, are the outcomes in some sub-populations independent of a given explanatory variable?
> Can you think of any inverted causal relationships in this sub-population?
> (I.e., for a given sub-population $B \rightarrow A$ instead of $A \rightarrow B$?)
> Consider adding these sub-population indices to one's initial causal graph, or if this is not clear enough, draw amended causal graphs for each sub-population.
> Now, one can actually relax.
> This concludes the ``purely mental'' drafting of one's causal graph.
> In the next section, we'll look at testing this graph against data, and making any edits deemed empirically necessary.

Reviewer: ```suggestion
After celebrating, take a moment to pursue the following graph editing exercises.
First, think about how the graph might differ across sub-populations.
What sub-populations, if any, exist in your population of interest?
Are there any causal relationships that should or should not not exist for a given sub-population?
For instance, are the outcomes in some sub-populations independent of a given explanatory variable?
Can you think of any inverted causal relationships specific to this sub-population?
(I.e., for a given sub-population, does $B \rightarrow A$ instead of $A \rightarrow B$?)
Consider adding these sub-population indices to one's initial causal graph, or if this is not clear enough, draw modified causal graphs for each sub-population of interest.
Now, one can actually relax.
This concludes the ``purely mental'' drafting of one's causal graph.
In the next section, we'll look at testing this graph against data, and making any edits deemed empirically necessary.
```

   \item remove any variables that could cause bias in our causal inferences,
   \item connect all variables in our graph according to our a-priori beliefs about causal relations amongst them
   \item consider how the graph structure may differ between individuals and subgroups within the population.
\end{enumerate}
The following paragraphs describe these steps in detail.

To begin, we adopt the position of a researcher concerned about population level relationships.
This means we will think through what is a likely generative model for all individuals.
Later in this section, we will devote time to thinking about how subgroups and individual heterogeneity may affect our causal graphs.

Now, we add our first variable(s) to our graph, the outcome variable(s) of interest in our problem.
Then, we list all the variables we believe to influence the outcome(s).
We refer to these variables as our initial explanatory variables.

Next, we iterate through these initial explanatory variables.
For each current explanatory variable in the iteration, we think of variables that may modify the effect of the current explanatory variable on the outcome(s) of interest.

> We refer to these variables as effect modifiers.

Reviewer 2: I think an example of an effect modifier would be extremely useful. As it stands they can be easily confused with confounders and/or mediators, which _modify_ the effect too.

Author: Added!

Note that some effect modifiers may be a part of our list of initial explanatory variables.
For any effect modifiers that we think of, outside of the list of initial explanatory variables, we add them to our causal graph.

After adding explanatory and effect modifying variables to the graph, we turn our attention to mediating variables.
A mediating variable is one through which an explanatory variable influences our outcome(s) of interest.
Again, we iterate through each explanatory variable.
On each iteration, we brainstorm all possible mediating variables.
For instance, consider how the presence of a bike lane influences bicycle mode choice.
We hypothesize that an individual's subjective perception of safety is the primary (or sole) mediator through which bicycle lane presence influences mode choice.

> Accordingly, we add subjective perception of safety to our causal graph for travel mode choice.

Reviewer 2: I think these steps are clear, but would benefit greatly from illustrative real world examples for each type of variables, and why it matters from a practical perspective to classify those variables into those _different_ categories and dealing with them. 

For example, why does it matter to add a mediator like perception of safety for a bike lane dummy since we don't observe it, and if it's truly the sole mediator for the bike lane effect, then what does it add to the analysis? The answers to those questions could be interesting and useful, but that's not very clear right now.

Author: Added examples for effect modifiers and references to the front-door criterion and surrogacy analysis for mediating variables.

In reality, how do we think our system evolves over time?
If we consider multiple observations of a given decision maker, how does that decision maker's observed variables at time $t$ partially cause future variables important to the context or outcome(s) for that decision maker at time $t' > t$?
How do the actions of a decision maker $i$ at time $t$ partially cause the future context or outcomes of a decision maker $j$?
We should add explicit nodes to our graph, subscripted or denoted by time, to show the cross-time causal relationships in our system.

At this point, we have added to our causal graph all the outcome, explanatory, effect modifying, mediating, confounding, and time-indexed variables that we believe are relevant for our problem.
However, they are all disconnected nodes, i.e., singletons in the graph.
We now focus on pruning nodes from this graph, before drawing our final hypothesized connections.
In particular, we focus on pruning ``post-outcome'' variables that are not part of the causal graph for future time periods or other observations.

> The reason for this is that conditioning on such post-outcome variables would bias our causal effect estimates.

Reviewer 2: This is interesting. I agree with you totally that conditioning on post-treatment variables is bad in practice if you're interested in the causal effect of a variable, but conditioning on such variables is so common in practice that trying to convince researchers to remove them from there models would cause a lot of resistance, specifically transportation demand models. 

For example, so many variables would be considered "post-treatment" variables if your treatment is a bike lane, e.g., bike ownership, residential choice, cars per household, etc. You will always have those variables in a model, because people are interested in them too. The value of a causal graph here is to keep track of those inter-variable dependencies when coming up with a generative model of the data and estimating structural equations. Even the example of distance -> time/cost could be framed that way if your treatment is "distance" as we discussed. This is where maybe we can rephrase the write-up on those variables to highlight those nuances? Let's discuss!

Author: Resolving as I think we discussed the difference was in your thinking post-treatment vs my thinking post-outcome.

To remove the problematic variables, we iterate through each of the non-outcome variables in our graph, and we assess whether each variable is actually a result of the outcome (perhaps in combination with other variables in our graph).
These post-outcome variables temporally follow the outcome variable(s) but do not cause variables in the causal graph for other observations.
We remove all such post-outcome variables from our graph.

Now is a good time to step back and consider what other researchers have thought about our problem.
Specifically, we should conduct a literature review to see how other researchers have conceptualized the topic that we are working on.
Have they included variables that we have not?
Were those variables related our outcomes of interest?
If so, should we add these variables to our causal graph? How should these variables enter our graph?

> Do the included variables of other researchers point suggest the existence of confounders in their work that we should include in our graph?

Reviewer 2: ```suggestion
Do the included variables of other researchers suggest the existence of confounders in their work that we should include in our graph?
```

Author: 🙏

Have other researchers ascribed differing roles to our graph's current variables than we have?

> For example, have other researchers judged a variable to be confounder, when we solely thought of the variable as an effect modifier?

Reviewer 2: ```suggestion
For example, have other researchers judged a variable to be a confounder, when we solely thought of the variable as an effect modifier?
```

Author: 🙏

> As we answer these questions, critically examine the evidence for these alternative decisions to see if we should also reconsider how we're judging our variables.

Reviewer 2: ```suggestion
As we answer these questions, we should critically examine the evidence for these alternative decisions to see if we should also reconsider how we're judging our variables.
```

Author: 🙏
